src/ucp/tag/rndv.c

```diff
@@ -468,66 +455,40 @@ static void ucp_rndv_req_send_rtr(ucp_request_t *rndv_req, ucp_request_t *rreq,
     ucp_request_send(rndv_req, 0);
 }
 
-static void ucp_rndv_get_lanes_count(ucp_request_t *rndv_req)
-{
-    ucp_ep_h ep        = rndv_req->send.ep;
-    ucp_lane_map_t map = 0;
-    uct_rkey_t uct_rkey;
-    ucp_lane_index_t lane;
-
-    if (ucs_likely(rndv_req->send.rndv_get.lane_count != 0)) {
-        return; /* already resolved */
-    }
-
-    while ((lane = ucp_rndv_req_get_zcopy_rma_lane(rndv_req, map, &uct_rkey))
-            != UCP_NULL_LANE) {
-        rndv_req->send.rndv_get.lane_count++;
-        map |= UCS_BIT(lane);
-    }
-
-    rndv_req->send.rndv_get.lane_count = ucs_min(rndv_req->send.rndv_get.lane_count,
-                                                 ep->worker->context->config.ext.max_rndv_lanes);
-}
-
 static ucp_lane_index_t
 ucp_rndv_get_zcopy_get_lane(ucp_request_t *rndv_req, uct_rkey_t *uct_rkey)
 {
-    ucp_lane_index_t lane;
-
-    lane = ucp_rndv_req_get_zcopy_rma_lane(rndv_req,
-                                           rndv_req->send.rndv_get.lanes_map,
-                                           uct_rkey);
+    ucp_lane_index_t lane_idx;
+    ucp_ep_config_t *ep_config;
+    ucp_rkey_h rkey;
+    uint8_t rkey_index;
 
-    if ((lane == UCP_NULL_LANE) && (rndv_req->send.rndv_get.lanes_map != 0)) {
-        /* lanes_map != 0 - no more lanes (but BW lanes are exist because map
-         * is not NULL - we found at least one lane on previous iteration).
-         * reset used lanes map to NULL and iterate it again */
-        rndv_req->send.rndv_get.lanes_map = 0;
-        lane = ucp_rndv_req_get_zcopy_rma_lane(rndv_req,
-                                               rndv_req->send.rndv_get.lanes_map,
-                                               uct_rkey);
+    if (ucs_unlikely(!rndv_req->send.rndv_get.lanes_map_all)) {
+        return UCP_NULL_LANE;
     }
 
-    return lane;
+    lane_idx   = ucs_ffs64_safe(rndv_req->send.rndv_get.lanes_map_used);
```

Comment:
maybe just have 'rndv_req->send.rndv_get.curr_lane_idx' instead of 'rndv_req->send.rndv_get.lanes_map_used'?
the index will go from 0 to rndv_req->send.rndv_get.lanes_count - 1

Reply:
discussed offline